### DISCUSSION TITLE
Skip handlers and reset only a few selected ones?

### DISCUSSION THREAD

**Discussion by Geddard** [mswjs/msw#759] [Q&A]
Hi there, I'm checking the Docs and some issues here trying to figure out if there's a way of doing two things:

On real time:
1) signal the MSW somehow to say "for this endpoint, stop responding with mocks", i can think of some complicated ways of handling this myself from the setup point of view of my app, but i would love to have some flag i can pass to the handler for example that can be based of a store state, to say "skip this one" for example

2) After i used `worker.use()` to override an endpoint, is there some way to just restore one overriden endpoint?
Same as before, i could save in some state the current list of overrides, reset everything and re-apply the ones i want to keep overriden, but if could do something like `worker.resetHandlers(...identifiers)`, or something similar, that would be awesome

Ideas?

Thanks and kind regards!

**Comment by kettanaito** [comment]
Hey, @Geddard. Thanks for reaching out. 

> signal the MSW somehow to say "for this endpoint, stop responding with mocks"

You can create an override that returns nothing from the resolver:

```js
it('stop mocking /user', () => {
  server.use(
    rest.get('/user', () => undefined)
  )
})
```

The way requests are handled, is that if there's a matching handler, it will be used to produce a mocked response. If the handler returns nothing (either explicitly or implicitly), it's a signal for MSW to stop handler lookup and bypass the request.

This is a good use case to prioritize #676  and introduce a new `res.forward()` function to explicitly forward (bypass) a request.

> is there some way to just restore one overridden endpoint?

It's a great question. `.resetHandlers()` removes _all_ overrides that were not present when the original worker/server instance was created, so you cannot use it to restore one particular override. 

If you wish to use a request handler override just once, consider using [res.once()](https://mswjs.io/docs/api/response/once) in the override:

```js
server.use(
  rest.get('/user', (req, res, ctx) => res.once(ctx.json({ firstName: 'John' }))))
)
```

Once this override triggers, it will deactivate itself, and the next matching handler will be used (which can be another override also). 

There is no option to programmatically restore a single override at the moment. Please, could you provide more info about your use case where you need to restore just one override? We could learn from it and come up with a suitable API, if applicable.

**Reply by Geddard** [reply]
Your first suggestion doesn't work, the MSW still responds with the mocks.
The only thing that has worked for me is using the patch response pattern, so i keep a list of the endpoints and match it with a checkbox list state like this

```
const bypassMockedResponse = useCallback(async (req, res, ctx) => {
    const originalResponse = await ctx.fetch(req)
    const originalResponseData = await originalResponse.json()

    return res(ctx.json(originalResponseData))
  }, [])

  const handleToggleEndpoint = (checked, page, url) => {
    checked ? worker.resetHandlers() : worker.use(rest.get(pages[page][url], bypassMockedResponse))

    setCheckedItems({ ...checkedItems, [url]: checked })
  }
  ```

But I guess the `.forward()` would be much nicer in this case. Looking forward to that feature :)

As for the resetHandlers, I'm planning on using this same list to pass to resetHandlers a list of the original handlers with the overridden ones on top that are not checked back when I only check back one.

**Reply by kettanaito** [reply]
What version of `msw` do you have?

```sh
$ npm ls msw
```

**Reply by Geddard** [reply]
> What version of `msw` do you have?
> 
> ```shell
> $ npm ls msw
> ```

0.27.1

**Reply by kettanaito** [reply]
Could you please update to the latest version? #416 was added relatively recently, that's why my first suggestion didn't work for you—the logic is not present in the older versions. `npm i msw@latest` should help.

**Reply by Geddard** [reply]
> Could you please update to the latest version? #416 was added relatively recently, that's why my first suggestion didn't work for you—the logic is not present in the older versions. `npm i msw@latest` should help.

I just tried and i still get serviceWorker when i try to use with an undefined return :/

```
const handleToggleEndpoint = (checked, page, url) => {
    checked ? worker.resetHandlers() : worker.use(rest.get(pages[page][url], () => undefined))

    setCheckedItems({ ...checkedItems, [url]: checked })
  }
```

**Comment by dagadbm** [comment]
I have another use case for skipping the selected ones.

I have some unit tests and some helper functions that mock a particular set of requests (which the redux store does)

However, I would like to be able to, sometimes, override these mock requests with other.

It would be nice to have some resetHandlers API that receives a string (which is the matched url) and removes the interceptor from it.
